### ISSUE TITLE
Screenreader accessibility: NVDA review cursor commands don't work in most of tree views and list views.

### ISSUE THREAD

**Issue by mltony** [microsoft/vscode#153995]
Issue Type: <b>Bug</b>

Does this issue occur when all extensions are disabled?: Yes

This issue is related to screenreader accessibility for the blind. It can only be reproduced with NVDA screenreader installed and running. I am using the latest version NVDA 2022.1.

## Steps to Reproduce:
1. Open any project in VSCode and press Control+Shift+E to go to explorer TreeView.
2. Press up/down keys to make sure some file is selected. These commands work finea and NVDA announces selection.
3. Press review cursor commands, such as Numpad 8 or Numpad 5 or Numpad 2  to read current line/word/character (make sure NumLock is off).

## Expected behavior: 
These commands should read current selection.

## Actual behavior: 
These commands read the word "blank" or nothing at all.

This issue seems to reproduce in all tree views and most list views in VSCode. A couple of other examples are:
* Watch variables in Debug pane.
* List of problems (Control+Shift+M)
* List of definitions and list of references (When pressing F12 or Alt+Shift+F12)

## Why this issue is important to visually-impaired users?
1. When there is only 1 item in the list view (like only a single variable in Watch view), there is no way to make screenreader to speak it since Up/Down commands don't do anything.
2. Often times the contents are too long (as in problems view). When reading an error I would often want to read it word-by-word or copy a single word to clipboard. Currently the only way appears to be to copy the entire message, paste it to an editor like Notepad and read by word there, which is more tedious than necessary.

## Investigation on my side
I am one of NVDA contributors and I performed an  investigation from NVDA's side to make sure that NVDA is not at fault here.
NVDA interacts with Chromium using IAccessible2 interface; it sees that all the items in VSCode views are presented as TreeViewItems. They appear to expose IAccessibleTextObject interface, but it typically returns an empty string consisting of a single whitespace. To reproduce this, you can focus on any tree view item, press Insert+Control+Z to open NVDA console and type:
```
>>> focus.IAccessibleTextObject.nCharacters
1
>>> focus.IAccessibleTextObject.text(0,1)
' '
```
I also tried to look at why Chromium presents empty text to NVDA for those TreeView items, but I didn't make much progress here since I didn't find an easy way to view HTML code representing those TreeView items.
I tried some toy examples from MDN, such as grid view: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/grid_role
And TreeItem role (I had to include tabIndex=0 attribute to make these items focusable): https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role
In both cases the problem doesn't reproduce, which makes me think that the issue must be in the way VSCode renders HTML for Chromium.

VS Code version: Code 1.68.1 (30d9c6cd9483b2cc586687151bcbcd635f373630, 2022-06-14T12:48:58.283Z)
OS version: Windows_NT x64 10.0.22000
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 4700U with Radeon Graphics          (8 x 1996)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.37GB (9.65GB free)|
|Process Argv|--crash-reporter-id 9a96d9f3-a840-44d8-9d62-0bac44c9d4b9|
|Screen Reader|yes|
|VM|0%|
</details><details><summary>Extensions (8)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.3.0
goto-next-previous-member|mis|0.0.6
python|ms-|2022.8.1
vscode-pylance|ms-|2022.6.30
jupyter|ms-|2022.5.1001601848
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.8
indent-nav|Ton|1.0.3

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscoreces:30445986
pythondataviewer:30285071
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vscscmwlcmt:30465135
cppdebug:30492333
vscaat:30438848
vsclangdc:30486549

```

</details>

**Comment by mltony** [comment]
cc: @isidorn (I know you're on vacation, so feel free to take a look when you're back)
Also I just wanted to mention that it's not a new regression, it has always been like this as long as I've been using VSCode - since 2-3 years ago.

**Comment by isidorn** [comment]
@mltony thanks for filling this issue and for a great investigation, and sorry for the slow response I was on vacation as you found out already :)

So the question is what does Chromium need in order to return the right content for the `focus.IAccessibleTextObject.text` call. For the lists we use a combination of `aria-activedescendant` and setting `aria-label` on items. Because of the `aria-activedescendant` the actual element that has focus is the actual list, not the element. So the gist of the problem seems to be that NVDA should respect the activedescendant when looking for the focused element, or that chromium should respect it when returning the focused element.

@joanmarie might Orca also support getting text from the focused element?

Let's see if somebody has good ideas, if not I can file an issue on the Chromium side to get their thoughts.

Also assigning to @joaomoreno as the list widget owner

**Comment by mltony** [comment]
I can see that currently focused element has role of TreeViewItem, not TreeView, so I suspect that NVDA knows how to properly deal with active descendants:
```
>>> focus.role
<Role.TREEVIEWITEM: 22>
>>> focus.parent.role
<Role.SECTION: 85>
>>> focus.parent.parent.role
<Role.TREEVIEW: 21>
```
SO IAccessibleTextObject.text function is called apparently on the treeVIewItem and it returns a single space character instead of actual text. FWIW, I also just noticed that `focus.name` returns correct name and internally this calls:
```
>>> f.IAccessibleObject.accName(f.IAccessibleChildID)
'api.py'
```
So apparently, there are two ways to communicate text content of an item through IAccessible2 API, and only one of them seems to be working. 
I am checking my toy example that I copied from MDN: https://pastebin.com/2eNgTGa5
In this example we have a table instead of treeView, but I hope that text should be handled in a similar manner by Chromium and NVDA. And I see that both ways (both IAccessible2 API calls) return correct text in this table. Moreover, setting aria-label doesn't seem to have any effect on either API call.
@isidorn, would it be possible to somehow produce an example of HTML code that VSCode is generating for those tables and tree views? Then if I could reproduce the problem in Chrome browser, this would allow us to easily bisect and figure out who'd be the owner of this problem. And hopefully find a solution.

**Comment by brunoprietog** [comment]
What I do is to go to the child object and then to the sibling. For example, press control+shift+E, select a file with the arrows and then with NVDA+shift+down arrow and then right arrow you can focus on the object containing the folder name. Then with NVDA and arrows you can read normally. It is certainly not ideal but it is possible

**Comment by mltony** [comment]
@brunoprietog, Ah, you're right, if I examine the hierarchy of objects, I indeed see that every treeViewItem has two StaticText children, and the second one indeed contains text I'm looking for.
So with this in mind, the problem can probably be tackled on NVDA's side. I'll close this issue for now as indeed VSCode presents necessary information. I will try to hack NVDA to consume it properly.

**Comment by isidorn** [comment]
@brunoprietog thanks for jumping in.
@mltony let me know how it goes on the NVDA side and if we should do some changes on the VS Code side to help here. Thanks a lot for your effort, I really appreciate it!
